### PR TITLE
fix: ensure env manager role has permission to upload

### PR DESCRIPTION
**Pull request by Lou1415926**
Basically bringing back logic in #3562  and #3567 .

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/cli/deploy/patch/env.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+// EnvironmentPatcher checks if the environment needs a patch and perform the patch when necessary.
+type EnvironmentPatcher struct {
+	Env             *config.Environment
+	TemplatePatcher environmentTemplateUpdateGetter
+}
+
+// EnsureManagerRoleIsAllowedToUpload checks if the environment manager role has the necessary permission to upload 
+// objects and patch the permissions if not.
+func (d *EnvironmentPatcher) EnsureManagerRoleIsAllowedToUpload(bucketName string) error {
```
Comment (dannyrandall) [inline internal/pkg/cli/deploy/patch/env.go]:
```suggestion
// EnsureManagerRoleIsAllowedToUpload checks if the environment manager role has the necessary permissions to upload 
// objects to bucket and patches the permissions if not.
func (d *EnvironmentPatcher) EnsureManagerRoleIsAllowedToUpload(bucket string) error {
```

File: internal/pkg/cli/deploy/patch/env.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	}
+
+	var tpl Template
+	if err := yaml.Unmarshal([]byte(body), &tpl); err != nil {
+		return fmt.Errorf("unmarshal environment template to find EnvironmentManagerRole policy statement: %v", err)
+	}
+	if len(tpl.Resources.ManagerRole.Properties.Policies) == 0 {
+		return errors.New("unable to find policies for the EnvironmentManagerRole")
+	}
+	// lines and columns are 1-indexed, so we have to substract one from each.
```
Comment (dannyrandall) [inline internal/pkg/cli/deploy/patch/env.go]:
```suggestion
	// lines and columns are 1-indexed, so we have to subtract one from each.
```

File: internal/pkg/cli/deploy/patch/env.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	linesBefore := lines[:statementLineIndex]
+	linesAfter := lines[statementLineIndex:]
+	updatedLines := append(linesBefore, append(strings.Split(permissions, "\n"), linesAfter...)...)
+	updatedBody := strings.Join(updatedLines, "\n")
+
+	// Update the Environment template with the new content.
+	// CloudFormation is the only entity that's allowed to update the EnvManagerRole so we have to go through this route.
+	// See #3556.
+	var errEmptyChangeSet *cloudformation.ErrChangeSetEmpty
+	// o.prog.Start("Update the environment's manager role with permission to upload artifacts to S3")
```
Comment (dannyrandall) [inline internal/pkg/cli/deploy/patch/env.go]: wanna just delete these lines?
```suggestion
```

**Review by dannyrandall — APPROVED** [review]
looks great!

File: internal/pkg/cli/deploy/patch/env.go (hunk 3 above)
Comment (Lou1415926) [reply internal/pkg/cli/deploy/patch/env.go]: Forgot to add the client for `prog`! I uncommented these lines.

File: internal/pkg/cli/deploy/patch/env.go [hunk 4]
```diff
@@ -95,13 +102,13 @@ func (d *EnvironmentPatcher) grantManagerRolePermissionToUpload(app, env, execRo
 	// CloudFormation is the only entity that's allowed to update the EnvManagerRole so we have to go through this route.
 	// See #3556.
 	var errEmptyChangeSet *cloudformation.ErrChangeSetEmpty
-	// o.prog.Start("Update the environment's manager role with permission to upload artifacts to S3")
+	d.Prog.Start("Update the environment's manager role with permission to upload artifacts to S3")
```
Comment (huanjani) [inline internal/pkg/cli/deploy/patch/env.go]: just curious why `EnvironmentPatcher` is `d`...? :)

Comment (huanjani) [reply internal/pkg/cli/deploy/patch/env.go]: oh HA. nvm. just saw the next commit.

Comment (Lou1415926) [reply internal/pkg/cli/deploy/patch/env.go]: they've been renamed to `p` in the latest commit! [b5165b7](https://github.com/aws/copilot-cli/pull/3956/commits/b5165b798711655f00d47effb04234eb6d2f01be)

**Review by huanjani — APPROVED** [review]
LGTM!

**Pull request merged**